src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs

```diff
@@ -297,6 +299,352 @@ public virtual void CreateCollation<T>(string name, T state, Func<T, string, str
+            var numberOfArguments = 1;
+            if (function != null)
+            {
+                if (typeof(T1).UnwrapNullableType().UnwrapEnumType() == typeof(object[]))
+                {
+                    stepFunction = (ctx, _, args) =>
+                    {
+                        if (ctx.state == null)
+                        {
+                            ctx.state = new StateValue<TContext> { Value = seed };
+                        }
+
+                        ((StateValue<TContext>)ctx.state).Value = function(((StateValue<TContext>)ctx.state).Value, (T1)GetArgumentsArray(args));
+                    };
+
+                    numberOfArguments = -1;
+                }
+                else
+                {
+                    stepFunction = (ctx, _, args) =>
+                    {
+                        if (args.Length < 1)
+                        {
+                            throw new ArgumentOutOfRangeException(); // TODO: add error message
+                        }
+
+                        if (ctx.state == null)
+                        {
+                            ctx.state = new StateValue<TContext> { Value = seed };
+                        }
+
+                        ((StateValue<TContext>)ctx.state).Value = function(((StateValue<TContext>)ctx.state).Value, GetArgument<T1>(args[0]));
+                    };
+                }
+
+                finalFunction = (ctx, _) =>
+                    SetFunctionResult(resultSelector(((StateValue<TContext>)ctx.state).Value), ctx);
+            }
+
+            CreateAggregateFunction(
+                name, numberOfArguments, stepFunction, finalFunction);
+            return resultSelector(seed);
+        }
+
+        private void CreateAggregateFunction(
+            string name,
+            int numberOfArguments,
+            delegate_function_aggregate_step stepFunction,
+            delegate_function_aggregate_final finalFunction,
+            [CallerMemberName] string callerName = "")
+        {
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+            if (State != ConnectionState.Open)
+            {
+                throw new InvalidOperationException(Resources.CallRequiresOpenConnection(callerName));
+            }
+
+            var rc = raw.sqlite3_create_function(_db, name, numberOfArguments, null, stepFunction, finalFunction);
+            SqliteException.ThrowExceptionForRC(rc, _db);
+        }
+
+        private T GetArgument<T>(sqlite3_value value)
+        {
+            var type = typeof(T).UnwrapNullableType().UnwrapEnumType();
+            if (type == typeof(DBNull))
+            {
+                return (T)(object)DBNull.Value;
+            }
+            if (type == typeof(object))
+            {
+                return (T)GetArgumentObject(value);
+            }
+
+            return SqliteDataReader.GetValue<T>(
```

Comment:
I tried the use of a class inheriting from DbDataRecord, but a large number of functions which have to be implemented for the IDataRecord interface, depend on the state of the DataReader, e.g.
```[csharp]
        public override bool IsDBNull(int ordinal)
            => _closed
                ? throw new InvalidOperationException(Resources.DataReaderClosed(nameof(IsDBNull)))
                : !_stepped || _done
                    ? throw new InvalidOperationException(Resources.NoData)
                    : GetSqliteType(ordinal) == raw.SQLITE_NULL;
```
I therefore created a new static class with all the value conversion.